### ISSUE TITLE
RFC: Support Server Side Rendering / Dynamic Routes in Eleventy

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#1727]
After a demo of running Eleventy inside of a serverless function (Netlify Function) https://github.com/11ty/demo-eleventy-cloud it makes sense to talk about supporting this as a feature in Eleventy (either supporting in core or via hooks for plugins).

This would enable a lot of very cool use cases while still keeping the zero clientside JavaScript ethos of Eleventy intact.

For example, I have this example working:

Build time generated template:
<img width="510" alt="image" src="https://user-images.githubusercontent.com/39355/113490802-29a8da80-9492-11eb-80c2-6102a7a26df9.png">

Build time generated template:
<img width="355" alt="image" src="https://user-images.githubusercontent.com/39355/113490811-375e6000-9492-11eb-9ac8-693b48cfe656.png">

Request time generated template:
<img width="373" alt="image" src="https://user-images.githubusercontent.com/39355/113490817-3f1e0480-9492-11eb-96ea-aecbbee7847c.png">

Request time generated template:
<img width="506" alt="image" src="https://user-images.githubusercontent.com/39355/113490821-4513e580-9492-11eb-8fb0-5e7d43c3f9ef.png">

The only difference between these two templates is `eleventyMode: "static"` or `eleventyMode: "dynamic"` in front matter.

There is some more work before this is production ready but the single data property API is exciting!

**Comment by Jaza** [comment]
Interesting. Although it goes against what has been the definition of 11ty from day one: a static site generator. Now it's also a dynamic server-side renderer!

Anyway, apparently 11ty is able to be both things, without any hackery as such required. So I guess that's cool.

Why would I want to use 11ty for this, when there are so many other mature frameworks around? (e.g. fastify, koa).

**Comment by zachleat** [comment]
Eleventy will continue to be a static site generator, of course. Although you could probably use this for your whole site, I would _not_ recommend it. This is intended to complement Eleventy, and will most likely end up as an Eleventy plugin because it will be coupled to specific host configurations to work in a serverless environment.

Primarily this will facilitate DX improvements to the Eleventy Cloud work that’s been happening to run Eleventy in a serverless function:

* https://www.zachleat.com/web/netlify-powers-events/
* https://twitter.com/zachleat/status/1365091172138569730
* https://github.com/11ty/demo-eleventy-cloud/ This Eleventy Cloud demo is fine, but I’d like to make it even easier to use.

The primary use cases I’m thinking of right now:

1. You want to expose a few URLs to have live-ish previews in your CMS.
2. Most of your build is static, but maybe you want a few dynamic routes for authenticated content that differs for each user.
3. (Placeholder for a third use case I can’t talk about yet 😅)

**Comment by solution-loisir** [comment]
So this would facilitate the use of the new [on-demand builder](https://www.netlify.com/blog/2021/04/14/faster-builds-for-large-sites-on-netlify-with-on-demand-builders-now-in-early-access/) feature from Netlify. I think it's brilliant to solve incremental builds with incremental rendering. It also eases serverless rendering of dynamic routes (without the use of on-demand builder) which can be an interesting solution in some cases. Overall, I feel this is leading 11ty in the right direction.

**Comment by solution-loisir** [comment]
If the `eleventyMode` property could be computed either directly or through `eleventyComputed` it would allow to switch from static to dynamic from say a CMS or through a time based automation. The use case that I'm thinking for that would be subscriptions to courses. During subscriptions period, when you need to serve fresh data about availability you use dynamic and revert to static the rest of the time. I thinking about something like that:

In course.json:
```json
{
  "subscriptions": true
}
```
In template.njk:
```twig
---
eleventyComputed: 
  eleventyMode: "{{ 'dynamic' if course.subscriptions else 'static' }}"
---
```

**Comment by Ryuno-Ki** [comment]
I can uncover 4.!

Winning back @andy-piccalilli from NextJS: https://css-tricks.com/give-your-eleventy-site-superpowers-with-environment-variables/ taken up a notch :D

**Comment by zachleat** [comment]
Working demo of this here: https://twitter.com/zachleat/status/1387101402279907334

**Comment by cfjedimaster** [comment]
I believe I have a use case similar to https://github.com/11ty/eleventy/issues/1727#issuecomment-821463314. My blog has posts from 2003 to now. All posts are markdown files and I use folders based on year, month, and date. As I already uses tags for something else, I create my collection via a glob. 

What I'd like to do is make a *subset* of that collection load on demand, something like: 2003/*, 2004/*, etc. Every few years I'd up that a bit such that only the last ten years or so are built at build time.

**Comment by robdodson** [comment]
> Not sure about this one quite yet but you might also imagine a local developer experience that generated templates at request time for a shorter/faster feedback loop during development.

I'm also pretty interested in this use case. When I'm doing local development, I often only care about one or two pages needing to be rebuilt. When we deploy our site, if it takes a while to build the whole thing, I don't really care. Also we're on Firebase Hosting, so we can't take advantage of incremental builders anyway.

But for local dev I would _love_ to be able to just rebuild a single page by toggling eleventyMode for my entire site, or for major sections of the site using eleventyComputed.

**Comment by PythonCreator27** [comment]
Will this support all types of platforms (Vercel functions, netlify functions, begin, cloudflare, etc.) and become platform agnostic (like [SvelteKit](https://github.com/sveltejs/kit) and its adapters)? Or will this remain constrained to Netlify functions alone?

**Comment by zachleat** [comment]
One of the goals of this project is to keep the plugin as provider-independent as possible, though the first release will use Netlify! I will likely need help customizing for alternate providers, but will provide the hooks necessary to do so. @dephiros @PythonCreator27

**Comment by zachleat** [comment]
https://github.com/11ty/demo-eleventy-serverless has been updated to use this and this is now available in the Eleventy 1.0 canary 29

Official docs (on 11ty.dev) coming next
